File: src/saliency.cpp
#include <opencv2/opencv.hpp>

int main()
{
    cv::Mat img = cv::imread("test.jpg");
    cv::cvtColor(img, img, CV_BGR2GRAY);
    float ratio = 128.0 / img.cols;
    cv::resize(img, img, cv::Size(img.cols*ratio, img.rows*ratio));

    cv::Mat planes[] = {cv::Mat_<float>(img), cv::Mat::zeros(img.size(), CV_32F)};
    cv::Mat complexImg;
    cv::merge(planes, 2, complexImg);
    cv::dft(complexImg, complexImg);
    cv::split(complexImg, planes);

    cv::Mat mag, logmag, smooth, spectralResidual;
    cv::magnitude(planes[0], planes[1], mag);
    cv::log(mag, logmag);
    cv::boxFilter(logmag, smooth, -1, cv::Size(3,3));
    cv::subtract(logmag, smooth, spectralResidual);
    cv::exp(spectralResidual, spectralResidual);

    planes[0] = planes[0].mul(spectralResidual) / mag;
    planes[1] = planes[1].mul(spectralResidual) / mag;

    cv::merge(planes, 2, complexImg);
    cv::dft(complexImg, complexImg, cv::DFT_INVERSE | cv::DFT_SCALE);
    cv::split(complexImg, planes);
    cv::magnitude(planes[0], planes[1], mag);
    cv::multiply(mag, mag, mag);
    cv::GaussianBlur(mag, mag, cv::Size(9,9), 2.5, 2.5);
    cv::normalize(mag, mag, 1.0, 0.0, NORM_MINMAX);

    cv::imshow("Saliency Map", mag);
    cv::waitKey(0);
}

File: src/saliency.py
import numpy as np
import cv2

WIDTH = 128   # has a great influence on the result

if __name__ == '__main__':
	img = cv2.imread('test.jpg', 0)
	img = cv2.resize(img, (WIDTH,WIDTH*img.shape[0]/img.shape[1]))

	c = cv2.dft(np.float32(img), flags = cv2.DFT_COMPLEX_OUTPUT)
	mag = np.sqrt(c[:,:,0]**2 + c[:,:,1]**2)
	spectralResidual = np.exp(np.log(mag) - cv2.boxFilter(np.log(mag), -1, (3,3)))

	c[:,:,0] = c[:,:,0] * spectralResidual / mag
	c[:,:,1] = c[:,:,1] * spectralResidual / mag
	c = cv2.dft(c, flags = (cv2.DFT_INVERSE | cv2.DFT_SCALE))
	mag = c[:,:,0]**2 + c[:,:,1]**2
	cv2.normalize(cv2.GaussianBlur(mag,(9,9),3,3), mag, 0., 1., cv2.NORM_MINMAX)

	cv2.imshow('Saliency Map', mag)
	c = cv2.waitKey(0) & 0xFF
	if(c==27 or c==ord('q')):
		cv2.destroyAllWindows()
